noir-projects/noir-contracts/contracts/nft_contract/src/test/transfer_to_private.nr

```diff
@@ -0,0 +1,92 @@
+use crate::test::utils;
+use dep::aztec::{
+    hash::pedersen_hash, keys::getters::get_current_public_keys, prelude::{AztecAddress, NoteHeader},
+    oracle::unsafe_rand::unsafe_rand, protocol_types::storage::map::derive_storage_slot_in_map
+};
+use crate::{types::nft_note::NFTNote, NFT};
+
+#[test]
+unconstrained fn transfer_to_private_to_self() {
+    // The transfer to private to self is done in `utils::setup_mint_and_transfer_to_private` and for this reason
+    // in this test we just call it and check the outcome.
+
+    // Setup without account contracts. We are not using authwits here, so dummy accounts are enough
+    let (env, nft_contract_address, user, _, token_id) = utils::setup_mint_and_transfer_to_private(/* with_account_contracts */ false);
+
+    // User should have the note in their private nfts
+    utils::assert_owns_private_nft(nft_contract_address, user, token_id);
+
+    // Since the NFT was sent to private, the public owner should be zero address
+    utils::assert_owns_public_nft(env, nft_contract_address, AztecAddress::zero(), token_id);
+}
+
+#[test]
+unconstrained fn transfer_to_private_to_a_different_account() {
+    // Setup without account contracts. We are not using authwits here, so dummy accounts are enough
+    let (env, nft_contract_address, sender, recipient, token_id) = utils::setup_and_mint(/* with_account_contracts */ false);
+
+    let note_randomness = unsafe_rand();
+    let transient_storage_slot_randomness = unsafe_rand();
+    // Sender will be the msg_sender/transfer_preparer in prepare_transfer_to_private
+    let transfer_preparer_storage_slot_commitment = pedersen_hash(
+        [sender.to_field(), transient_storage_slot_randomness],
+        NFT::TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+    );
+
+    // We prepare the transfer
+    let prepare_transfer_to_private_call_interface = NFT::at(nft_contract_address).prepare_transfer_to_private(
+        sender,
+        recipient,
+        note_randomness,
+        transient_storage_slot_randomness
+    );
+    env.call_private_void(prepare_transfer_to_private_call_interface);
+
+    // Finalize the transfer of the NFT
+    let finalize_transfer_to_private_call_interface = NFT::at(nft_contract_address).finalize_transfer_to_private(token_id, transfer_preparer_storage_slot_commitment);
+    env.call_public(finalize_transfer_to_private_call_interface);
+
+    // Store the finalized note in the cache
+    let mut context = env.private();
+    let recipient_npk_m_hash = get_current_public_keys(&mut context, recipient).npk_m.hash();
+    let private_nfts_recipient_slot = derive_storage_slot_in_map(NFT::storage().private_nfts.slot, recipient);
+
+    env.store_note_in_cache(
+        &mut NFTNote { token_id, npk_m_hash: recipient_npk_m_hash, randomness: note_randomness, header: NoteHeader::empty() },
+        private_nfts_recipient_slot,
+        nft_contract_address
+    );
```

Comment:
This is simply manually adding a note, right? Do you know why this is called 'cache'?

Reply:
This is a bit of an "implementation detail leak". It's called like that because it calls the `notifyCreatedNote` oracle and then it gets stored in `ExecutionNoteCache` within the TXEOracle. So it doesn't use the nore processor flow but instead it gets stored in cache which is usually used for not-yet delivered notes.

Agree that it's a bad name as the dev does not need to care about it. Will rename it in a followup PR.

Reply:
Tackled in https://github.com/AztecProtocol/aztec-packages/pull/8547